Biddle Verbs: Queues the Most Expensive Verbs for the Next Tick if the Server Is Overloaded

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This pr goes through: /client/Click(), /client/Topic(), /mob/living/verb/resist(), /mob/verb/quick_equip(), /mob/verb/examinate(), and /mob/verb/mode() and makes them queue their functionality to a subsystem to execute in the next tick if the server is overloaded. To do this a new subsystem is made to handle most verbs called SSverb_manager, if the server is overloaded the verb queues itself in the subsystem and returns, then near the start of the next tick that verb is resumed with the provided callback. The verbs are called directly after SSinput, and the subsystem does not yield until its queue is completely finished.

The exception are clicks from player input since they are extremely important for the feeling of responsiveness. I considered not queuing them but theyre too expensive not to, suffering from a death of a thousand cuts performance wise from many many things in the process adding up. Instead clicks are executed at the very start of the next tick, as the first action that SSinput completes, before player movement is processed even.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
A few months ago, before I died I was trying to figure out why games at midpop (40-50 people) had non zero and consistent time dilation without maptick being consistently above 28% (which is when the MC stops yielding for maptick if its overloaded). I found it out, started working on this pr, then promptly died. luckily im a bit less dead now

the current MC has a problem: the cost of verbs is completely and totally invisible to it, it cannot account for them. Why is this bad? because verbs are the last thing to execute in the tick, after the MC and SendMaps have finished executing. 
![tick diagram2](https://user-images.githubusercontent.com/15794172/159179946-f4afaeef-7192-44a3-94a1-556e4c854c42.png)
If the MC is overloaded and uses 100% of the time it allots itself this means that if SendMaps uses the amount its expected to take, verbs have at most 2% of the tick to execute in before they are overtiming and thus delaying the start of the next tick. This is bad, and im 99% sure this is the majority of our overtime. 

Take Click() for example. Click isnt listed as a verb but since its called as a result of client commands its executed at the end of the tick like other verbs. in this random 80 pop sybil round profile i had saved on my computer [sybil 80 pop (2).txt](https://github.com/tgstation/tgstation/files/8312060/sybil.80.pop.2.txt) /client/Click() has an overtime of only 1.8 seconds, which isnt that bad. however it has a self cpu of 2.5 seconds meaning 1.8/2.5 = 72% of its time is overtiming, and it also is calling 80.2 seconds worth of total cpu, which means that more than 57.7 seconds of overtime is attributed to just /client/Click() executing at the very end of a tick. the reason why this isnt obvious is just because the verbs themselves typically dont have high enough self cpu to get high enough on the rankings of overtiming procs to be noticed, all of their overtime is distributed among a ton of procs they call in the chain. 

Since i cant guarantee the MC resumes at the very start of the next tick due to other sleeping procs almost always resuming first: I time the duration between clicks being queued up for the next tick and when theyre actually executed. if it exceeds 20 milliseconds of added latency (less than one tenth the average human reaction time) clicks will execute immediately instead of queuing, this should make instances where a player can notice the added latency a vanishingly small minority of cases. still, this should be tm'd
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
offloads clicks to SSinput if the server is overloaded, and adds stats

makes SSinput process queued clicks before processing movement because input sensitivity is easier to tell for clicks than it is for movement

Kylerace committed:
adds first parts of SSverb_manager

Kylerace committed:
makes SSinput clicks use a new verb_callback type

Kylerace committed:
finishes docs and queuing logic on SSverb_manager

Kylerace committed:
makes /client/Topic queue window updates and datum/Topic calls

Kylerace committed:
makes examinate(), mode(), quick_equip(), and resist() use verb queuing

Kylerace committed:
makes SSinput only pass queued clicks if a usr was recorded

optimumtact commented:
![definition of biddle within the context of tgstation, which is when the meter increases and things are unlocked](https://user-images.githubusercontent.com/2544323/159184174-cfdc210e-57cb-4d5e-ba86-3c632c03b2fd.png)

Kylerace commented:
must be tm'd first

Time-Green commented:
You’re completely insane

👍

LemonInTheDark commented:
Many of these paths rely on usr having a valid value. What will we do about them?

Kylerace commented:
> Many of these paths rely on usr having a valid value. What will we do about them?

ideally throw a runtime error, since these should be replacing verbs usr should always exist

MrStonedOne commented:
> Many of these paths rely on usr having a valid value. What will we do about them?

This uses callbacks which preserve the usr at creation time when executing. (usr is just a global var byond has a tendency to set, its not read only) the code to handle preserving usr is highly admin fuckary resistant, and shouldn't present a security issue.

MrStonedOne commented on code/modules/client/client_procs.dm:
```diff
@@ -121,8 +121,17 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 
+	//fun fact: Topic() acts like a verb and is executed at the end of the tick like other verbs. So we have to queue it if the server is
+	//overloaded
+	if(hsrc && TRY_QUEUE_VERB(src, .proc/Topic_wrapper, hsrc, href, href_list))
+		return
```
exempt admin topics
```suggestion
	if(hsrc && hsrc != holder && TRY_QUEUE_VERB(src, .proc/Topic_wrapper, hsrc, href, href_list))
		return
```

Kylerace committed:
Update code/modules/client/client_procs.dm

Co-authored-by: Kyle Spier-Swenson <kyleshome@gmail.com>

LemonInTheDark commented:
> > Many of these paths rely on usr having a valid value. What will we do about them?
> 
> This uses callbacks which preserve the usr at creation time when executing. (usr is just a global var byond has a tendency to set, its not read only) the code to handle preserving usr is highly admin fuckary resistant, and shouldn't present a security issue.

oh we preserve vars like that in callbacks? that's based

stylemistake commented on code/modules/tgui/external.dm:
```diff
@@ -213,5 +213,9 @@
 		payload = json_decode(href_list["payload"])
 	// Pass message to window
 	if(window)
+		//this proc is called from /client/Topic() which is called at the very end of a tick like a verb, so try to queue this high cost proc if
+		//the sever is overloaded.
+		if(TRY_QUEUE_VERB(window, /datum/tgui_window.proc/on_message, type, payload, href_list))
+			return TRUE
```
Might potentially worsen the current state of tgui, because window initialization is time sensitive. It's better to move TRY_QUEUE_VERB towards ui_act part, since that is a less critical path and can be the most cpu intensive.

stylemistake replied:
https://github.com/tgstation/tgstation/blob/master/code/modules/tgui/tgui.dm#L311

Kylerace committed:
does style's review: moves verb yielding from tgui_window to /tgui

Kylerace committed:
makes verb/pointed use the queuing systen

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into biddle_verbs

Kylerace committed:
fixes merge conflict in mob/living/pointed

Kylerace committed:
makes it compile

Kylerace commented:
it doesnt blow everything up but i need to investigate some problems

Kylerace committed:
fixes successful click queuing always causing double clicks

Kylerace committed:
fixes tgui not working at all :trollface:

stylemistake committed:
Slightly better naming for "ui_act" callback

Kylerace committed:
Merge remote-tracking branch 'origin/biddle_verbs' into biddle_verbs

Kylerace committed:
fixes conflict, and makes both queues return false if not initialized

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into biddle_verbs

Kylerace committed:
adds runlevels check to both verb queuing procs

Kylerace committed:
adds new verb queue defines and a var that tracks maptick on master

LemonInTheDark commented:
Verbs executing after map info is sent has been a massive boon hasn't it

Kylerace committed:
makes verb/set_tab() use verb queuing

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -34,6 +52,55 @@ SUBSYSTEM_DEF(input)
+	if(control != "mapwindow.map" || !ismob(usr) || QDELING(usr) || QDELETED(clicked_atom))
+		return FALSE
+
+	//high priority because clicks should be as low latency as possible, deferring to the beginning of the next SSinput run should rarely delay
+	//a click by more than a few milliseconds, but since the MC almost always resumes after every other sleeping proc this isnt guaranteed
```
How is it possible for SSinput to delay a click? it should be fired immediately as the next tick starts shouldn't it?

Kylerace replied:
not always because the MC loop isnt the only sleeping proc that exists, and most other sleeping procs will run before Master in the tick theyre scheduled to start because Master almost always sleeps for 1 tick and other procs usually sleep for longer which puts them further in the queue.

its totally possible for a ton of continually sleeping, expensive procs to delay the MC resuming a relevant amount. if that happens we stop queuing clicks until the delay averages down below the threshold again (though i should make the average MC_AVG_FAST_UP_SLOW_DOWN())

LemonInTheDark replied:
Ah I see, the old meteor behavior

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -34,6 +52,55 @@ SUBSYSTEM_DEF(input)
+
+	//high priority because clicks should be as low latency as possible, deferring to the beginning of the next SSinput run should rarely delay
+	//a click by more than a few milliseconds, but since the MC almost always resumes after every other sleeping proc this isnt guaranteed
+	if(!TICK_CHECK_HIGH_PRIORITY \
+		|| average_click_delay >= MAXIMUM_CLICK_LATENCY \
+		|| FOR_ADMINS_IF_CLICKS_BROKE_immediately_execute_all_clicks \
+		|| !initialized \
+		|| !(runlevels & Master.current_runlevel))
```
Assertion, clicks from admins should never ever be delayed

Kylerace replied:
done

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,23 @@
+#define IS_IN_VERB_TICK_PORTION (world.time >= Master.last_run && MAPTICK_LAST_INTERNAL_TICK_USAGE != Master.last_maptick)
+
+///attempt to queue this verb process if the server is overloaded. evaluates to FALSE if queuing isnt necessary or if it failed.
+#define TRY_QUEUE_VERB(_object, _proc, _tick_check, _args...) (IS_IN_VERB_TICK_PORTION && (TICK_USAGE > _tick_check) && SSverb_manager.initialized && SSverb_manager.queue_verb(_object, _proc, _args))
```
same here

Kylerace replied:
done

Kylerace committed:
makes admins not get their verbs or clicks queued

Kylerace committed:
makes average_click_delay use MC_AVG_FAST_UP_SLOW_DOWN

Kylerace committed:
Merge branch 'master' into biddle_verbs

Kylerace committed:
REEEEEEEEEEEEEEEE

LemonInTheDark commented on code/__DEFINES/time.dm:
```diff
@@ -46,6 +46,10 @@ When using time2text(), please use "DDD" to find the weekday. Refrain from using
 #define SATURDAY "Sat"
 #define SUNDAY "Sun"
 
+#define MILLISECONDS * 0.01
```
```suggestion
#define MILLISECONDS *0.01
```
*gags*

Kylerace replied:
:walter:

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,23 @@
+///default queuing tick_usage threshold for most verbs which can allow a small amount of latency to be processed in the next tick
+#define VERB_DEFAULT_QUEUE_THRESHOLD 85
+
+///client verbs execute after Master and after SendMaps, so we're in that portion iff we agree with Master on world.time and disagree on maptick,
```
```suggestion
///client verbs execute after Master and after SendMaps, so we're in that portion if we agree with Master on world.time and disagree on maptick,
```

Kylerace replied:
iff = if and only if

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,23 @@
+//its for this reason why player clicks are handled in SSinput before even movement - semantic latency could cause someone to move out of range
+//when the verb finally processes but it was in range if the verb had processed immediately and overtimed.
+
+///queuing tick_usage threshold for verbs that are high enough priority that they only queue if the server is overtiming. only use for critical verbs
```
```suggestion
///queuing tick_usage threshold for verbs that are high enough priority that they only queue if the server is overtiming. 
///ONLY use for critical verbs
```

LemonInTheDark commented on code/controllers/master.dm:
```diff
@@ -32,6 +32,8 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 	/// world.time of last fire, for tracking lag outside of the mc
 	var/last_run
+	/// the last MAPTICK_LAST_INTERNAL_TICK_USAGE the mc recorded, used for tracking tick portions outside the MC.
+	var/last_maptick
```
this should always be 0, no?

Kylerace replied:
what?

LemonInTheDark replied:
idk for some reason I thought world.map_cpu reset at the start of the tick
seems weird to store like this regardless tho idk

Kylerace replied:
It's used to check if a given proc is executing after maptick but honestly it's not needed

Kylerace replied:
Also it's not reset

LemonInTheDark replied:
ye I uh, I realized that once I thought about it for more then 2 seconds

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -9,6 +9,24 @@ SUBSYSTEM_DEF(input)
 
 	var/list/macro_set
 
+	///what mobs are currently queued to click an atom. ran through every tick before movement input is processed.
+	///list verb callbacks storing information for clicks
+	var/list/datum/callback/verb_callback/queued_clicks = list()
+	///running average of how many clicks inputted by a player the server processes every second. used for the subsystem stat entry
+	var/clicks_per_second = 0
+	///count of how many clicks onto atoms have elapsed before being cleared by fire(). used to average with clicks_per_second.
+	var/current_clicks = 0
+	///acts like clicks_per_second but only counts the clicks actually processed by SSinput itself while clicks_per_second counts all clicks
+	var/delayed_clicks_per_second = 0
+	///running average of how many movement iterations from player input the server processes every second. used for the subsystem stat entry
+	var/movements_per_second = 0
+	///running average of the amount of real time clicks take to truly execute after the command is originally sent to the server.
+	///if a click isnt delayed at all then it counts as 0 deciseconds.
+	var/average_click_delay = 0
+
+	///if for whatever reason clicks arent being executed fast enough set this to TRUE and all clicks will immediately execute
+	var/FOR_ADMINS_IF_CLICKS_BROKE_immediately_execute_all_clicks = FALSE
```
Why do this in the input subsystem exactly? cause it's on brand?

Kylerace replied:
Clicks are the only client input I can think of where latency matters more than resolving movement

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -35,6 +53,56 @@ SUBSYSTEM_DEF(input)
+	queued_clicks += VERB_CALLBACK(src, .proc/wrap_Click, clicked_atom, location, control, params)
+	return TRUE
+
+///stupid workaround for byond not recognizing the /atom.proc/Click typepath for the queued click callbacks
```
did you try /atom/Click or /atom/proc/Click?
using . here would i think fuck you up

Kylerace replied:
I think I tried both

Kylerace replied:
doesnt work with /atom/Click or /atom/proc/Click

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -35,6 +53,56 @@ SUBSYSTEM_DEF(input)
+		current_clicks++
+		deferred_clicks_this_run++
+
+	queued_clicks.Cut() //is ran all the way through every run, no exceptions
```
is there any cost to this?

Kylerace replied:
To the Cut()? It happens once per tick

LemonInTheDark replied:
idk man my brain sees list operation and goes brrrr, it's fine

LemonInTheDark commented on code/controllers/subsystem/statpanel.dm:
```diff
@@ -264,7 +264,7 @@ SUBSYSTEM_DEF(statpanels)
 
 	stat_tab = tab
-	SSstatpanels.immediate_send_stat_data(src)
+	QUEUE_OR_CALL_VERB(SSstatpanels, /datum/controller/subsystem/statpanels:proc/immediate_send_stat_data, VERB_DEFAULT_QUEUE_THRESHOLD, src)
```
:?

Kylerace replied:
good catch, no reason for that

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+	wait = 1
+	flags = SS_TICKER
+	priority = FIRE_PRIORITY_DELAYED_VERBS
+	runlevels = ALL//trollface
```
```suggestion
	runlevels = ALL //trollface
```
help my brain is broken

Kylerace replied:
so is mine, now it uses RUNLEVEL_DEFAULT | RUNLEVEL_INIT

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+
+	///list of callbacks to procs called from verbs or verblike procs that were executed when the server was overloaded and had to delay to the next tick.
+	///this list is ran through every tick, and the subsystem does not yield until this queue is finished.
+	var/list/verb_queue = list()
```
is this list typed?

Kylerace replied:
Yes, containing /datum/callback/verb or whatever it is

LemonInTheDark replied:
should its typepath reflect that?

Kylerace replied:
sure

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+	///running average of how many verb callbacks are executed every second. used for the stat entry
+	var/verbs_executed_per_second = 0
+
+	var/FOR_ADMINS_IF_VERBS_FUCKED_immediately_execute_all_verbs = FALSE
```
we should give them a debug verb that sets this and the say and input one

Kylerace replied:
Good idea

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+ * like SSinput can start at the same time they were supposed to, leading to a smoother experience for the player since ticks arent riddled with
+ * minor hangs over and over again.
+ */
+SUBSYSTEM_DEF(verb_manager)
```
It would be useful to emphasize the verblike proc aspect of this, so ui acts and such

Kylerace replied:
True

Kylerace replied:
uh, what name would work for that? verblike_manager doesnt have quite the same ring to it. the portion of the tick its supposed to work with is called the verb portion anyways

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+
+	var/FOR_ADMINS_IF_VERBS_FUCKED_immediately_execute_all_verbs = FALSE
+
+///queue a callback on a proc meant to resume processing for a verb for the next tick. returns TRUE if it was queued, FALSE otherwise.
```
this description is kinda hard to parse. maybe it'd be better to describe the intent rather then the step by step function?

Kylerace replied:
done'd

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+/datum/controller/subsystem/verb_manager/proc/queue_verb(datum/thing_to_call, proc_to_call, ...)
+	//since this queue is meant specifically to delay user input as little as possible without overrunning the tick, the callback must have usr.
+	//no, you cant queue everything up for this eat shit. if its not user input then it cant go into a queue that never yields
+	if(QDELETED(thing_to_call) \
+	|| !proc_to_call \
+	|| !ismob(usr) \
+	|| QDELING(usr) \
+	|| usr.client?.holder /*dont worry admins cant cause overtime :clueless:*/ \
+	|| FOR_ADMINS_IF_VERBS_FUCKED_immediately_execute_all_verbs \
+	|| !initialized \
+	|| !(runlevels & Master.current_runlevel))
+		return FALSE
```
this check isn't too expensive is it?

Kylerace replied:
verbs arent expensive because thousands of them happen per tick and they add up, theyre expensive because a few verbs take on the order of milliseconds or few hundred microseconds and they happen a few times per tick (at the end). so this should never be a relevant cost

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,86 @@
+/datum/controller/subsystem/verb_manager/fire(resumed)
+	var/executed_verbs = 0
+
+	for(var/datum/callback/verb_callback as anything in verb_queue)
+		if(!verb_callback)
+			continue
```
how would a null get here? could we just do += length(verb_queue)?
Or is the accuracy important

Kylerace replied:
It probably can't without a very weird bug, but we can't allow for a runtime to drop an entire ticks verb execution

LemonInTheDark replied:
so why not drop the as anything and let byond to the typechecking?

Kylerace replied:
true, i probably intended to stack_trace() there or smth

LemonInTheDark commented on code/modules/client/client_procs.dm:
```diff
@@ -121,8 +121,17 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
+		return
 	..() //redirect to hsrc.Topic()
 
+///dumb workaround because byond doesnt seem to recognize the .proc/Topic() typepath for /datum/proc/Topic() from the client Topic,
+///so we cant queue it without this
```
see click question, .proc will break here even if it wasn't internal

Kylerace replied:
see original review answer

LemonInTheDark commented on code/modules/client/client_procs.dm:
```diff
@@ -912,14 +926,21 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			to_chat(src, span_danger("Your previous click was ignored because you've done too many in a second"))
 			return
 
+	//check if the server is overloaded and if it is then queue up the click for next tick
+	if(SSinput.queue_click_from_usr(object, location, control, params))
+		return
```
Clicks are likelyish to read to icon2htmls, which can be quite expensive. I guess this check will never pass in high maptick cases, but what about right below that line?

Kylerace replied:
if its below the threshold then it continues the click as normal

LemonInTheDark commented on code/modules/keybindings/bindings_atom.dm:
```diff
@@ -22,3 +22,5 @@
 	else
 		user?.Move(get_step(src, movement_dir), movement_dir)
+
+	return !!movement_dir //true if there was actually any player input
```
recorded input and doing the actually expensive part of the move are different things, no?

Kylerace replied:
This is just to record subsystem stats for number of actual client movement inputs per second

LemonInTheDark commented on code/modules/mob/dead/observer/observer.dm:
```diff
@@ -673,14 +673,11 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 	target.faction = list("neutral")
 	return TRUE
 
-//this is a mob verb instead of atom for performance reasons
-//see /mob/verb/examinate() in mob.dm for more info
-//overridden here and in /mob/living for different point span classes and sanity checks
```
rip comment

Kylerace committed:
reviews for potato, removes last_maptick because its useless

Kylerace committed:
first steps towards making SSinput and SSspeech_controller subtypes

i want them all to be subtypes of SSverb_manager so that they dont have the same operations and methods repeated for each of them

Kylerace committed:
makes the queue defines work

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into biddle_verbs

Kylerace committed:
lemon reviews

Kylerace committed:
lol im dumb. fixes me and whisper verbs

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into biddle_verbs

Kylerace committed:
meninblackflash.gif

Kylerace committed:
Update code/__DEFINES/verb_manager.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

LemonInTheDark commented:
JERRYYYYYY

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into biddle_verbs

Kylerace committed:
why the fuck did the merge conflict do that

LemonInTheDark commented:
I'm going to merge this I swear to god

Kylerace committed:
i need to stop doing this shit

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,33 @@
+ * when the verb finally processes but it was in range if the verb had processed immediately and overtimed.
+ */
+
+///queuing tick_usage threshold for verbs that are high enough priority that they only queue if the server is overtiming. 
+///ONLY use for critical verbs
+#define VERB_OVERTIME_QUEUE_THRESHOLD 100
+///queuing tick_usage threshold for verbs that need lower latency more than most verbs.
+#define VERB_HIGH_PRIORITY_QUEUE_THRESHOLD 95
+///default queuing tick_usage threshold for most verbs which can allow a small amount of latency to be processed in the next tick
+#define VERB_DEFAULT_QUEUE_THRESHOLD 85
```
what do these numbers represent?

Kylerace replied:
the tick_usage threshold that causes verbs to be queued when exceeded. ie if a verb has the threshold value VERB_DEFAULT_QUEUE_THRESHOLD, then when its ran, it will be queued for the next tick if tick_usage exceeds 85

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,33 @@
+
+///attempt to queue this verb process if the server is overloaded. evaluates to FALSE if queuing isnt necessary or if it failed.
+///_verification_args... are only necessary if the verb_manager subsystem youre using checks them in can_queue_verb()
+///if you put anything in _verification_args that ISNT explicitely put in the can_queue_verb() override of the subsystem youre using,
+///it will runtime.
```
feel like we should unit test for this somehow

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,33 @@
+///_verification_args... are only necessary if the verb_manager subsystem youre using checks them in can_queue_verb()
+///if you put anything in _verification_args that ISNT explicitely put in the can_queue_verb() override of the subsystem youre using,
+///it will runtime.
+#define TRY_QUEUE_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args...) (_queue_verb(_verb_callback, _tick_check, _subsystem_to_use, _verification_args))
```
if verification args are optional shouldn't they use ##?

Kylerace replied:
it seems to kill the parser when i do that

LemonInTheDark replied:
k

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,33 @@
+///it will runtime.
+#define TRY_QUEUE_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args...) (_queue_verb(_verb_callback, _tick_check, _subsystem_to_use, _verification_args))
+///queue wrapper for TRY_QUEUE_VERB() when you want to call the proc if the server isnt overloaded enough to queue
+#define QUEUE_OR_CALL_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args...) if(!TRY_QUEUE_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args)) {_verb_callback:InvokeAsync()};
```
ditto, also just make this a proclike macro it's easier to read

LemonInTheDark commented on code/__DEFINES/verb_manager.dm:
```diff
@@ -0,0 +1,33 @@
+///queue wrapper for TRY_QUEUE_VERB() when you want to call the proc if the server isnt overloaded enough to queue
+#define QUEUE_OR_CALL_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args...) if(!TRY_QUEUE_VERB(_verb_callback, _tick_check, _subsystem_to_use, _verification_args)) {_verb_callback:InvokeAsync()};
+
+//goes straight to SSverb_manager with default tick threshold
+#define DEFAULT_TRY_QUEUE_VERB(_verb_callback, _verification_args...) (TRY_QUEUE_VERB(_verb_callback, VERB_DEFAULT_QUEUE_THRESHOLD, null, _verification_args))
+#define DEFAULT_QUEUE_OR_CALL_VERB(_verb_callback, _verification_args...) QUEUE_OR_CALL_VERB(_verb_callback, VERB_DEFAULT_QUEUE_THRESHOLD, null, _verification_args)
+
+//default tick threshold but nondefault subsystem
+#define TRY_QUEUE_VERB_FOR(_verb_callback, _subsystem_to_use, _verification_args...) (TRY_QUEUE_VERB(_verb_callback, VERB_DEFAULT_QUEUE_THRESHOLD, _subsystem_to_use, _verification_args))
+#define QUEUE_OR_CALL_VERB_FOR(_verb_callback, _subsystem_to_use, _verification_args...) QUEUE_OR_CALL_VERB(_verb_callback, VERB_DEFAULT_QUEUE_THRESHOLD, _subsystem_to_use, _verification_args)
```
ditto

LemonInTheDark commented on code/__HELPERS/turfs.dm:
```diff
@@ -207,6 +207,11 @@ Turf and target are separate in case you want to teleport some distance from a t
 	if(!istype(checked_atom))
 		return
 
+	//Find coordinates
+	var/turf/atom_turf = get_turf(checked_atom) //use checked_atom's turfs, as it's coords are the same as checked_atom's AND checked_atom's coords are lost if it is inside another atom
+	if(!atom_turf)
+		return null
```
why?

Kylerace replied:
cant remember i need to stop doing this. was probably for the early return

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -1,15 +1,28 @@
+	var/current_clicks = 0
+	///acts like clicks_per_second but only counts the clicks actually processed by SSinput itself while clicks_per_second counts all clicks
+	var/delayed_clicks_per_second = 0
+	///running average of how many movement iterations from player input the server processes every second. used for the subsystem stat entry
+	var/movements_per_second = 0
```
not sure I agree with this as a stat, seems potentially misleading

Kylerace replied:
whats misleading about it?

LemonInTheDark replied:
https://github.com/tgstation/tgstation/pull/65589#discussion_r916417710 mostly. players often complete inputs, what's interesting is the cost of moving itself, rather then the cost of holding down the button, no?

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -29,12 +42,53 @@ SUBSYSTEM_DEF(input)
-		user.focus?.keyLoop(user.client)
+/datum/controller/subsystem/verb_manager/input/can_queue_verb(datum/callback/verb_callback/incoming_callback, control)
+	//make sure the incoming verb is actually something we specifically want to handle
+	if(control != "mapwindow.map")
+		return FALSE
```
why?

Kylerace replied:
code assumes the existence of a clicked atom which is only when the click happened in the map window.

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -29,12 +42,53 @@ SUBSYSTEM_DEF(input)
+	return TRUE
+
+///stupid workaround for byond not recognizing the /atom/Click typepath for the queued click callbacks
+/atom/proc/wrap_Click(location, control, params)
```
```suggestion
/atom/proc/_Click(location, control, params)
```
thoughts? standard wrapping syntax, see sdql

Kylerace replied:
alright

LemonInTheDark commented on code/controllers/subsystem/input.dm:
```diff
@@ -29,12 +42,53 @@ SUBSYSTEM_DEF(input)
+	var/moves_this_run = 0
+	var/deferred_clicks_this_run = 0 //acts like current_clicks but doesnt count clicks that dont get processed by SSinput
+
+	for(var/datum/callback/verb_callback/queued_click in verb_queue)
+		average_click_delay = MC_AVG_FAST_UP_SLOW_DOWN(average_click_delay, (REALTIMEOFDAY - queued_click.creation_time) SECONDS)
+		queued_click.InvokeAsync()
+
+		current_clicks++
+		deferred_clicks_this_run++
+
```
```suggestion
	for(var/datum/callback/verb_callback/queued_click in verb_queue)
		average_click_delay = MC_AVG_FAST_UP_SLOW_DOWN(average_click_delay, (REALTIMEOFDAY - queued_click.creation_time) SECONDS)
		queued_click.InvokeAsync()

	current_clicks += length(verb_queue)
	deferred_clicks_this_run += length(verb_queue)

```

LemonInTheDark replied:
unless this is to filter nulls in which case uh, how

Kylerace replied:
it is meant to filter elements of verb_queue that arent iterated over. i changed the for loop to stack_trace() and continue if it goes over something that shouldnt be there.

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+ * SSverb_manager, a subsystem that runs every tick and runs through its entire queue without yielding like SSinput.
+ * this exists because of how the byond tick works and where user inputted verbs are put within it.
+ *
+ * The byond tick proceeds as follows:
+ * 1. procs sleeping via walk() are resumed (i dont know why these are first)
+ * 2. normal sleeping procs are resumed, in the order they went to sleep in the first place, this is where the MC wakes up and processes subsystems.
+ *	a consequence of this is that the MC almost never resumes before other sleeping procs, because it only goes to sleep for 1 tick 99% of the time
+ *	and 99% of procs either go to sleep for less time than the MC (which guarantees that they entered the sleep queue earlier when its time to wake up)
+ *	and/or were called synchronously from the MC's execution, almost all of the time the MC is the last sleeping proc to resume in any given tick.
+ *	This is good because it means the MC can account for the cost of previous resuming procs in the tick, and minimizes overtime.
+ * 3. control is passed to byond after all of our code's procs stop execution for this tick
+ * 4. a few small things happen in byond internals
+ * 5. SendMaps is called for this tick, which processes the game state for all clients connected to the game and handles sending them changes
+ * 	in appearances within their view range. This is expensive and takes up a significant portion of our tick, about 0.45% per connected player
+ * 	as of 3/20/2022. meaning that with 50 players, 22.5% of our tick is being used up by just SendMaps, after all of our code has stopped executing.
+ *	Thats only the average across all rounds, for most highpop rounds it can look like 0.6% of the tick per player, which is 30% for 50 players.
+ * 6. After SendMaps ends, client verbs sent to the server are executed, and its the last major step before the next tick begins.
+ *	During the course of the tick, a client can send a command to the server saying that they have executed any verb. The actual code defined
+ *	for that /verb/name() proc isnt executed until this point, and the way the MC is designed makes this especially likely to make verbs
+ *	"overrun" the bounds of the tick they executed in, stopping the other tick from starting and thus delaying the MC firing in that tick.
```
make a .md file explaining processing order, and embed The Image TM in it, no balls

Kylerace replied:
done'd

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+
+	///list of callbacks to procs called from verbs or verblike procs that were executed when the server was overloaded and had to delay to the next tick.
+	///this list is ran through every tick, and the subsystem does not yield until this queue is finished.
+	var/list/datum/callback/verb_callback/verb_queue = list()
```
should we enter maptick overrun to complete this list?

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+	var/use_default_stats = TRUE
+
+	///if TRUE this will... message admins every time a verb is queued to this subsystem for the next tick with stats.
+	///for obvious reasons dont make this be TRUE on the code level this is for admins to turn on
```
admins, or debuggers?

Kylerace replied:
anyone with a holder

LemonInTheDark replied:
I mean the purpose, is it for admin use? or debugging stuff

Kylerace replied:
i suppose debugging

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+ * returns TRUE if the queuing was successful, FALSE otherwise.
+ */
+/proc/_queue_verb(datum/callback/verb_callback/incoming_callback, tick_check, datum/controller/subsystem/verb_manager/subsystem_to_use, ...)
+	if(TICK_USAGE < tick_check \
+	|| QDELETED(incoming_callback) \
+	|| QDELETED(incoming_callback.object) \
+	|| !ismob(usr) \
+	|| QDELING(usr))
+		return FALSE
```
how hot is this code

LemonInTheDark replied:
ok so shortcutting qdeleting isn't really worth it (I think I was thinking about a ? not totally sure it's been too long

Kylerace replied:
true, also im not sure why i check if usr is a mob

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+	|| QDELING(usr))
+		return FALSE
+
+	subsystem_to_use = subsystem_to_use || SSverb_manager
```
why not as a default arg

Kylerace replied:
because i was looking at _addtimer when making this. i dont see why not

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+ * returns TRUE if the queuing was successful, FALSE otherwise.
+ */
+/datum/controller/subsystem/verb_manager/proc/queue_verb(datum/callback/verb_callback/incoming_callback)
+	. = FALSE //errored
```
do we actually need this?

Kylerace replied:
we dont _need_ it, but if for whatever reason the queuing doesnt work we should return FALSE so that the verb can execute immediately.

LemonInTheDark commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,134 @@
+	return TRUE
+
+/datum/controller/subsystem/verb_manager/fire(resumed)
+	var/executed_verbs = 0
+
+	for(var/datum/callback/verb_callback/verb_callback in verb_queue)
+		verb_callback.InvokeAsync()
+		executed_verbs++
+
```
```suggestion
	for(var/datum/callback/verb_callback/verb_callback in verb_queue)
		verb_callback.InvokeAsync()
		
	var/executed_verbs = length(verb_queue)
```
doesn't have null filtering, do we care about that? why?

LemonInTheDark commented on code/modules/client/client_procs.dm:
```diff
@@ -124,8 +124,17 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 
+///dumb workaround because byond doesnt seem to recognize the .proc/Topic() typepath for /datum/proc/Topic() from the client Topic,
+///so we cant queue it without this
+/client/proc/Topic_wrapper(datum/hsrc, href, list/href_list)
```
related to click, _ as wrapper syntax

LemonInTheDark commented on code/modules/keybindings/bindings_atom.dm:
```diff
@@ -22,3 +22,5 @@
 	else
 		user?.Move(get_step(src, movement_dir), movement_dir)
+
+	return !!movement_dir //true if there was actually any player input
```
still think this is measuring the wrong thing, specially as a moving average

Kylerace replied:
i suppose it would be more accurate to only return this if a Move() occured.

LemonInTheDark commented on code/modules/mob/mob.dm:
```diff
@@ -508,6 +508,10 @@
 	set name = "Examine"
 	set category = "IC"
 
+	DEFAULT_QUEUE_OR_CALL_VERB(VERB_CALLBACK(src, .proc/finish_examinate, examinify))
+
+/mob/proc/finish_examinate(atom/examinify)
```
run_examinate?

Kylerace replied:
sure

LemonInTheDark reviewed (changes requested):
btw do we still wrap statpanel sourced icon2htmls?

Kylerace commented on code/modules/client/client_procs.dm:
```diff
@@ -918,14 +932,22 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			clicklimiter[SECOND_COUNT] = 0
-		clicklimiter[SECOND_COUNT] += 1+(!!ab)
+
+		clicklimiter[SECOND_COUNT] += 1 + (!!ab)
```
yep https://github.com/tgstation/tgstation/pull/36829

LemonInTheDark replied:
had that smell

Kylerace commented on code/modules/keybindings/bindings_client.dm:
```diff
@@ -1,5 +1,5 @@
-// Clients aren't datums so we have to define these procs indpendently.
-// These verbs are called for all key press and release events
+/// Clients aren't datums so we have to define these procs indpendently.
+/// These verbs are called for all key press and release events
 /client/verb/keyDown(_key as text)
```
tru

Kylerace committed:
renames wrapping procs to the form _proc()

Kylerace committed:
default subsystem arg in _queue_verb()

Kylerace committed:
makes the verb_manager fire()s stack_trace() if they find a bad element

Kylerace committed:
fixes keyLoop() return and makes it not call procs if theres no dir

Kylerace committed:
keyLoop() only returns non zero if /client/Move() was called

Kylerace committed:
creates TICK_ORDER.md to document the byond tick structure

LemonInTheDark commented:
Ready?

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into biddle_verbs

Kylerace committed:
fixes pointing

Kylerace commented:
@LemonInTheDark yes

optimumtact commented:
bro it don't lint

Kylerace commented:
💀

Kylerace committed:
whoops forgot to save this deleted code

LemonInTheDark commented:
I HAVE RETURNED FROM CENTURIES OF SLUMBER, AND NOW THE UNIVERSE MUST BURN

Vallat commented on code/controllers/subsystem/verb_manager.dm:
```diff
@@ -0,0 +1,121 @@
+ * in TRY_QUEUE_VERB() and co.
+ */
+/datum/controller/subsystem/verb_manager/proc/can_queue_verb(datum/callback/verb_callback/incoming_callback)
+	if(usr.client?.holder && !can_queue_admin_verbs \
+	|| FOR_ADMINS_IF_VERBS_FUCKED_immediately_execute_all_verbs \
+	|| !initialized \
+	|| !(runlevels & Master.current_runlevel))
+		return FALSE
```
Am I missing something or this check will always return FALSE because this subsystem has flag `SS_NO_INIT`? 🤔

LemonInTheDark replied:
kylerrrrrrrrrrrrrrrrrrrrrrrrr

Kylerace replied:
luckily it appears that verbs were being queued for SSinput because it actually initializes, so we know the behavior isnt totally broken
